scipio/src/sys/uring.rs

```diff
@@ -944,7 +935,7 @@ impl Reactor {
         if should_sleep {
             // We are about to go to sleep. It's ok to sleep, but if there
             // is a timer set, we need to make sure we wake up to handle it.
-            if let Some(dur) = timer_expiration {
+            if let Some(dur) = preempt {
                 let mut src = self.timeout_src.borrow_mut();
                 lat_ring.cancel_preempt_timer(&mut src);
```

Comment:
@glommer what if `should_sleep &= wakers.is_empty();` will evaluate to false in original code? then the preemption timer would not be rearmed. is it fine? it is safe now when we always have `poll_io(preempt)` before calling `run_one_task` because poll_io will rearm the timer, but aside of this, is it fine that it can left preempt timer not rearmed?
Maybe it is fine to cancel preempt timer at all before parking (sleeping) and I can do else cancel here for spinning case?

Reply:
The preempt timer only needs to be rearmed if there is an active task queue.
If we arrive at `wait` with no timeout, that means that the Executor wants us to sleep because there are no more active task queues.

The only thing that can change that, is if we generated wakers while we were processing the cancellations and remaining pending flushes (or if we have anything in the poll ring, because the poll ring can't sleep).

The timer we rearm here, is no the preempt timer:

```rust
            // We are about to go to sleep. It's ok to sleep, but if there
            // is a timer set, we need to make sure we wake up to handle it.
            if let Some(dur) = timer_expiration {
```

Those are user timers. Let's say for instance you, as a user, asked for this:
```rust
Timer::new(Duration::from_secs(3600))
```

and then we go to sleep right after.


We can't sleep forever: we have to rearm the timer for 1h from now otherwise you will miss it.
The fact that the preempt timer path is reused here is just an implementation detail so we can coalesce those things.